CNCF Federated Governance Rebase (#83)
operator-framework/community: GOVERNANCE.md

- [Removing Projects](#removing-projects)

## Values

The Operator Framework and its leadership embrace the following values:

* Openness: Communication and decision-making happens in the open and is discoverable for future
    reference. As much as possible, all discussions and work take place in public
    forums and open repositories.

* Fairness: All stakeholders have the opportunity to provide feedback and submit
    contributions, which will be considered on their merits.

* Community over Product or Company: Sustaining and growing our community takes
    priority over shipping code or sponsors' organizational goals. Each
    contributor participates in the project as an individual.

* Inclusivity: We innovate through different perspectives and skill sets, which
    can only be accomplished in a welcoming and respectful environment.

* Participation: Responsibilities within the project are earned through
    participation, and there is a clear path up the contributor ladder into leadership
    positions.

## Mission

The mission of the Operator Framework community is:
* To collaborate on areas related to developing, packaging, deploying

>     and managing the lifecycle Kubernetes Operators

Reviewer: nit: grammar? Maybe `and managing the lifecycle of Kubernetes Operators`?

# Operator Framework Project Governance

The Operator Framework umbrella project is composed as a federation of
individual projects, some independent and some interdependent, each of

> which focuses on some aspect of the goals of umbrella project. Our

Reviewer 2: ```suggestion
which focuses on some aspect of the goals of the umbrella project. Our
```

* Ensuring that the subproject creates and publishes regular releases;
* Holding regular and open subproject-wide discussions on feature planning and issues;
* Regular review of project contributors for advancement on the Contributor Ladder;
* Following subproject governance and ensuring that the subproject
    functions according to the Operator Framework values.
* Making final decisions on subproject changes that involve controversial trade-offs;
* Responding to security compromise reports;
* Supporting the Code of Conduct within their subproject and referring violations
  to the Code of Conduct Committee.

Additionally, the project's Approvers Committee will select, by majority
vote, one representative to the Steering Committee. This representative
need not be a member of the subproject's Approvers Committee, and will
be replaced or renewed by the committee annually.

Should a member of the Approvers Committee cease being active in the subproject,
violate the Code Of Conduct, or need to be removed for some other reason, they
may be removed by a 2/3 majority vote of the other Committee members, or a
majority vote of the Steering Committee.

### Default Subproject Governance

> Subprojects without governing documents default to:

Reviewer 3: Needs info.

Committee members, or a vote of the Steering Committee.

#### Subproject Steering Committee Representative

Additionally, the Approver Committee will select, by an election
conducted by the Operator Framework Election officials, one
representative of the project to the Operator Framework Steering
Committee. This representative need not be a member of the project's
Maintainer Committee, and will be replaced or renewed by the Committee
annually.

#### Subproject Governance Ammendments

If the project members wish to update or supplement governance
documents, this will require a ⅔ majority  of project Maintainers
followed by a ⅔ majority approval from th e Operator Framework steering
committee.                                
                                          
## Steering Committee                     
                                          
The overall Operator Framework umbrella project is governed by a Steering
Committee, which is selected as follows:

* One Maintainer representative from each member subproject
* `7 - number_of_subprojects` General "Community Representatives", with a minimum of 3 or 4, whichever will result in an odd number of committee members.

> ### Steering Committee Duties

Reviewer 4: We decided to limit CNCF Maintainer status to just the Steering committee, right? Should have a blurb in here about representing OF to CNCF and related things like Kubecon

Author: added in latest commit

The Steering Committee is responsible for the following tasks, any of
which may be delegated to a person or group selected by the committee:

* Reviewing and deciding on new projects to add to Operator Framework
* Arbitrating inter-project disagreements
* Selecting the Code of Conduct Committee and ratifying CoC judgements
* Removing projects which have become inactive
* Acting on escalated security or code quality issues
* Resolving issues that individual projects are unable to resolve

> * Administering project infrastructure, intellectual property, and resources

Reviewer 5: Should we explicitly call out that the Steering Committee can delegate administration of project infrastructure (or other Steering Committee duties)?

   OF Election Officials and calculated by the Condorcet Method.
3. Following the announcement of the Subproject results, the
   Community Representatives are chosen by all Members of the Operator
   Framework. The vote will be conducted by the OF Election Officials
   and calculated by the Condorcet Method.

If individuals are nominated to be Subproject Representative, the Subproject elections will be calculated in a random order. No individual can
hold more than one seat on the Steering Committee.

### Steering Committee Bootstrapping

Upon ratification of this document, the 7 elected Chairs of the Operator
Framework will become the Steering Committee for a period of one year,
afterwards Steering Committee elections will proceed as outlined in this
document.

## Governance Amendments

The Operator Framework Community may adopt or amend its Governing
Documents according to the following process:

1. Propose a change by opening an issue in the community, and if there
   seems to be general agreement, open a pull request to this document.
2. The change should be announced via the mailing list along with an invitation for discussion on the Pull Request.

> 3. A minimum of 7 days must pass between the announcement and the vote. A longer period can be requested with approval from the steering committee. All maintainers, approvers, subproject owners, and the Steering committee are eligible to vote.

Author: 4. The vote must remain open for a minimum of 7 days, the Steering Committee may extend the period.

> 3. A minimum of 7 days must pass between the announcement and the vote. A longer period can be requested with approval from the steering committee. All maintainers, approvers, subproject owners, and the Steering committee are eligible to vote.

Reviewer 5: If a subproject is very conservative with maintainer/approver status, they will have less influence than another similarly sized subproject that gives every contributor maintainer/approver status.

It seems like our voting rules are setting up incentives for subprojects to "race to the bottom" in terms of doling out approver/maintainer to swing votes in their favor.

Is there a way to structure general votes to enable subprojects to have appropriate influence in elections regardless of their internal contributor ladder behavior?

Perhaps we just allow all org members to vote regardless of their position on a contributor ladder? And we just require $some quorum of org members?

Author: We discussed this one with TAG Contributor Strategy
 https://cloud-native.slack.com/archives/CT6CWS1JN/p1660759198667419?thread_ts=1660759198.667419&cid=CT6CWS1JN

"""
@jberkus to be clear, what you are suggesting is:
l
Governance Amendments are approved via a 2/3 majority of the Steering Committee. 

each subproject will have at least 1 voting rep based on steering committee membership
"""

Reviewer 6: Yes, although the rep count is covered elsewhere in the document.

> For a vote to pass:

Author: , all of the following must occur

* A quorum of 3 Steering Committee members is required.
* A 2/3 majority of the Steering Committee quorum must approve.

> * A majority of the voting quorum must approve.

Author: A majority of the Maintainers and Approvers must approve.

the world. To avoid fatigue, the Steering Committee will replace at
least 2 members of the CoC Committee each year. Members of
the committee do not need to be members of Operator Framework if they
have sufficient other expertise.

The CoC Committee will receive reports of conduct violations confidentially,
and will discuss them in closed meetings. If a report is determined to be a
violation, they will recommend action on it appropriate to the scale, nature,
and context of the violation, from requiring an apology, up to expulsion of an
individual from the project. In the event that a contributor is to be demoted
or expelled, the CoC Committee will forward this recommendation to the Steering
Committee who will ratify it in a closed meeting. Should a member of the
Steering Committee be the offender or the reporter, this recommendation will
instead be forwarded to the CNCF staff for final arbitration.

## Adding New Subprojects

During the monthly Steering meeting, any project member may recommend projects
to become part of Operator Framework. These projects should have the following
characteristics:

* Have a mission that aligns with the Operator Framework goals

> * Are appropriately licensed

Author: We discussed this in the office hour-- this is vague, and up to the Steering committee to determine.

* Governance structure is fully documented in a GOVERNANCE file
* Governance rules are modifiable by project leaders through a defined process.
* Embrace the Operator Framework and CNCF values
* Are under active development

> * Consist of high quality code and designs

Author: + Subproject has at least 3 Approvers (Experimental Subprojects must have at least 1)

Before submitting an application to the Steering Committee, the applying project
must hold an internal vote of all major contributors to join
Operator Framework. The Steering Committee will then review the

> application. If it is accepted, the Committee

Author: How many Steering Committee votes are required to accept?

Author: 2/3 for graduated 1/2 for experimental

will assign one person to assist the subproject in their integration.

In some cases, promising but incomplete projects may be accepted as Experimental
Subprojects. Such Experimental Subprojects will be considered part of
Operator Framework, but will be marked as Experimental on the website and in code
repositories, in order to inform users. Experimental Subproject Members are considered
Members of Operator Framework, but the s ubproject is not entitled to a representative on the
Steering Committee. Experimental Subproj ects can apply for full status by opening an issue, which will be reviewed by the Steering Committee.
                                         
## Removing Projects                     
                                         
In some cases, projects will become inac tive or unmaintainable, or wish to separate
from Operator Framework. Any Steering Committee member may propose removal of a project on

> these grounds, and Steering can confirm this with a majority vote.

Author: s/majority/ 2\/3/

- [Governance Ammendments](#governance-ammendments)
- [Code of Conduct Committee](#code-of-conduct-committee)
- [Adding New Subprojects](#adding-new-subprojects)
- [Removing Projects](#removing-projects)

## Values

The Operator Framework and its leadership embrace the following values:

* Openness: Communication and decision-making happens in the open and is discoverable for future
    reference. As much as possible, all discussions and work take place in public
    forums and open repositories.

* Fairness: All stakeholders have the opportunity to provide feedback and submit
    contributions, which will be considered on their merits.

* Community over Product or Company: Sustaining and growing our community takes
    priority over shipping code or sponsors' organizational goals. Each
    contributor participates in the project as an individual.

* Inclusivity: We innovate through different perspectives and skill sets, which
    can only be accomplished in a welcoming and respectful environment.

* Participation: Responsibilities within the project are earned through
    participation, and there is a clear path up the contributor ladder into leadership

>     positions.

Author: +consensus seeking

Author: when discussed in the office hour, it was determined that consensus seeking is part of being inclusive

## Mission

The mission of the Operator Framework community is:
* To collaborate on areas related to developing, packaging, deploying
    and managing the lifecycle of Kubernetes Operators

* To develop informational resources that give the community an
    understanding of best practices, trade-offs, and value adds as it
    relates to developing, packaging, deploying and managing the
    lifecycle of Kubernetes Operators

* To identify suitable projects and gaps in the ecosystem for future
    collaboration and coordination with those projects.

### Scope

The Operator Framework community focuses on the following
developer-related topics of creating Kubernetes Operators:
* Code scaffolding to create Kubernetes Operators
* Libraries supporting common needs of Operator authors
* Packaging of Kubernetes Operators
* Testing of Kubernetes Operators

The Operator Framework community focuses on the following admin-related
topics of running Kubernetes Operators on a Kubernetes cluster:

> * Maintaining catalogs of Operators

Author: maintain tooling of catalogs

* Discovering and installing Operators from those catalogs

> * Managing the lifecycle of continuous updates and eventual removal of Operators

Author: Managing the lifecycle of Operators

> The Operator Framework community will work on developing best practices,

Author: The of community develops best practices, fosters collaboration, and improves tool interoperability

### Default Subproject Governance

Subprojects without governing documents default to:

#### Subproject Approvers Committee Responsibilities

All active Approvers of the project, as defined in the Operator
Framework Contributor Ladder, are members of the Maintainer Committee,
which governs that project. The project's Approver Committee is
responsible for the following project governance activities:

* Ensuring that the project creates and publishes regular releases;
* Holding regular, project-wide discussions on issues and planning for the project;
* Monthly review of project contributors for advancement on the Contributor Ladder;
* Making final decisions on project changes that involve controversial trade-offs;
* Responding to security compromise reports;
* Supporting the Code of Conduct within their project and referring violations to the Code of Conduct Committee.

Should a member of the Approver Committee cease being active in the
project, violate the Code Of Conduct, or need to be removed for some
other reason, they may be removed by a 2/3 majority vote of the other
Committee members, or a vote of the Steering Committee.

#### Subproject Steering Committee Representative

> Additionally, the Approver Committee will select, by an election

Reviewer 4: Add a blurb here explicitly saying if the subproject only puts forward a single nominee, then don't bother holding an election.

conducted by the Operator Framework Election officials, one
representative of the project to the Operator Framework Steering
Committee. This representative need not be a member of the project's
Approver Committee, and will be replaced or renewed by the Committee
annually.

#### Subproject Governance Ammendments

If a subproject wishes to update or supplement governance
documents, this will require a 2/3 majority of project Approvers
followed by a 2/3 majority approval from the Operator Framework steering

> committee.

Author: lets remove the 2/3 of approvers

## Steering Committee

The overall Operator Framework umbrella project is governed by a Steering
Committee, which is selected as follows:

* One Maintainer representative from each member subproject

> * `7 - number_of_subprojects` General "Community Representatives", with a minimum of 3 or 4, whichever will result in an odd number of committee members.

Reviewer 5: ```suggestion
* `7 - number_of_subprojects` OR 3 general "community representatives", whichever is greater.
```

If number of subprojects is >= 7, we still want community reps, so the minimum number of reps is 3.
Also, we don't need to worry about even/odd concerns because:
- we don't have very many simple majority votes
- even if we do, votes typically don't garner 100% turnout (i.e. 10 of 11 votes might vote and split 5/5).

Reviewer 5: Also Josh suggested we add a table to show some examples.

Author: I removed the algebra, should be clearer now.

a preference election online, using the Condorcet method. The Election
Officers will announce the selected candidates at the next regular
community meeting.

#### Steering Committee Election Procedure

1. Nominations for Operator Framework Representatives (Subproject Rep
   and Community Rep) will be collected for at least 14 days. Interested
   individuals should self-nominate, and they may be considered for
   Subproject Representative and Community Representative.
2. For each Subproject, one Subproject Representative is chosen by the
   Subproject's Approvers. Each Subproject vote will be conducted by the
   OF Election Officials and calculated by the Condorcet Method.
3. Following the announcement of the Subproject results, the
   Community Representatives are chosen by all Members of the Operator
   Framework. The vote will be conducted by the OF Election Officials
   and calculated by the Condorcet Method.

If individuals are nominated to be Subproject Representative, the Subproject elections will be calculated in a random order. No individual can
hold more than one seat on the Steering Committee.

### Steering Committee Bootstrapping

> Upon ratification of this document, the 7 elected Chairs of the Operator

Reviewer 7: This is not really clear to me, so the current 7 chairs will be steering committee for the next year?

Author: yes, thats correct. FWIW, if we adopt a new Subproject, they will also get a rep, so we would have 8 or more

Framework will become the Steering Committee for a period of one year,
afterwards Steering Committee elections will proceed as outlined in this
document.

## Governance Amendments

The Operator Framework Community may adopt or amend its Governing

> Documents according to the following process:

Author: include ladder

These individuals will be chosen based on their community management and
code of conduct experience, with diverse representation across the
committee, including employer, gender, race, background, and region of
the world. To avoid fatigue, the Steering Committee will replace at
least 2 members of the CoC Committee each year. Members of
the committee do not need to be members of Operator Framework if they
have sufficient other expertise.

The CoC Committee will receive reports of conduct violations confidentially,
and will discuss them in closed meetings. If a report is determined to be a
violation, they will recommend action on it appropriate to the scale, nature,
and context of the violation, from requiring an apology, up to expulsion of an
individual from the project. In the event that a contributor is to be demoted
or expelled, the CoC Committee will forward this recommendation to the Steering
Committee who will ratify it in a closed meeting. Should a member of the
Steering Committee be the offender or the reporter, this recommendation will
instead be forwarded to the CNCF staff for final arbitration.

## Adding New Subprojects

> 

Author: when a new subproject is added (and they gain a steering committee seat) This seat will be additive, (no existing community chairs will be removed, even if there are fewer at large seats next time)

During the monthly Steering meeting, any project member may recommend projects
to become part of Operator Framework. These projects should have the following
characteristics:

* Have a mission that aligns with the Operator Framework goals
* Are appropriately licensed
* Governance structure is fully documented in a GOVERNANCE file
* Governance rules are modifiable by project leaders through a defined process.
* Embrace the Operator Framework and CNCF values
* Are under active development
* Consist of high quality code and designs
* Have at least 3 Approvers (Experimental Subprojects must have at least 1)

Before submitting an application to the Steering Committee, the applying project

> must hold an internal vote of all major contributors to join

Author: ```suggestion
must hold an internal vote of Approvers to join
```

Author: discretion of the steering committee is implied since they will be voting.

Operator Framework. The Steering Committee will then review and vote on the

> application. A quorum of 3 Steering Committee members must vote, and

Reviewer 6: Rather than having quorums, my suggestion is that you simply require all SC votes to be as a % of all existing Steering seats.  That is, if there's 7 commitee members then 4 are required for any majority vote and 5 for any 2/3 vote, regardless of how that vote is taken.  Quorums, especially minority quorums, must have rules around notice periods etc. that can be annoying to administer.

accepted Subprojects require 2/3 of the quorum. Experimental subprojects
only require 1/2 of the quorum to approve. If the application is
accepted, the Committee will assign one person to assist the subproject
in their integration.

In some cases, promising but incomplete projects may be accepted as
Experimental Subprojects. Such Experimental Subprojects will be
considered part of Operator Framework, but will be marked as
Experimental on the website and in code repositories, in order to inform
users. Experimental Subproject Members are considered Members of
Operator Framework, but the subproject is not entitled to a
representative on the Steering Committee. Experimental Subprojects can
apply for full status by opening an issue, which will be reviewed by the
Steering Committee.

## Removing Projects

In some cases, projects will become inactive or unmaintainable, or wish to separate
from Operator Framework. Any Steering Committee member may propose removal of a project on
these grounds, and Steering can confirm this with a 2/3 vote.

> 
> Projects which still have contributors will then be moved to a repository in their
> own namespace. Projects which have ceased all activity are moved to an archived namespace.

Author: ```suggestion
```
Lets leave this to the discretion of the steering committee

other reason, they may be removed by a 2/3 majority vote of the other
Committee members, or a vote of the Steering Committee.

#### Subproject Steering Committee Representative

Additionally, the Approver Committee will select, by an election
conducted by the Operator Framework Election officials, one
representative of the project to the Operator Framework Steering
Committee. This representative need not be a member of the project's
Approver Committee, and will be replaced or renewed by the Committee
annually.

#### Subproject Governance Ammendments

> If a subproject wishes to update or supplement governance
> documents, this will require a 2/3 majority of project Approvers
> followed by a 2/3 majority approval from the Operator Framework steering
> committee.

Reviewer 6: This feels like a recipe to make it impossible to ever change governance.  Consider that the #1 reason to change written governance is that the prior rules have become inoperable, usuaully due to contributor departures.  Suggest instead:

```suggestion
If a subproject wishes to update or supplement governance
documents, this will require a 2/3 majority of project Approvers
or 2/3 majority approval from the Operator Framework steering
committee.
```

   Framework. The vote will be conducted by the OF Election Officials
   and calculated by the Condorcet Method.

If individuals are nominated to be Subproject Representative, the
Subproject elections will be calculated in a random order. No individual
can hold more than one seat on the Steering Committee.

### Steering Committee Bootstrapping

Upon ratification of this document, the 7 elected Chairs of the Operator
Framework will become the Steering Committee for a period of one year,
afterwards Steering Committee elections will proceed as outlined in this
document. 

## Governance Amendments

The Operator Framework Community may adopt or amend its Governance

> Docuements (`GOVERNANCE.md` and `contributor-ladder.md`) according to

Author: cue: spellcheck

1. Propose a change by opening an issue in the community, and if there
   seems to be general agreement, open a pull request to this document.
2. The change should be announced via the mailing list along with an invitation for discussion on the Pull Request.
3. A minimum of 7 days must pass between the announcement and the vote,
   and the vote must remain open for a minimum of 7 days. The Steering
   Committee may extend both periods.

For a vote to pass, all of the following must occur:
* A quorum of 3 Steering Committee members is required.

> * A 2/3 majority of the Steering Committee quorum must approve.

Reviewer 6: There's issues with quorums; will discuss during the meeting.
